File: fixtures/py-sample/app/validation.py
"""Invoice validation rules."""

from app.models import CustomerStatus, Invoice, User


class ValidationError(Exception):
    pass


def validate_invoice(invoice: Invoice, user: User):
    """Suspended or closed customers must never be billed."""
    if user.status != CustomerStatus.ACTIVE:
        raise ValidationError(f"customer {user.id} is {user.status.value}")
    if invoice.amount_cents <= 0:
        raise ValidationError("invoice amount must be positive")

File: fixtures/ts-sample/src/orders/validation.ts
import type { Order } from './model';
import { CustomerStatus, type User } from '../users/model';

export class OrderValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'OrderValidationError';
  }
}

/** Orders from suspended or closed customers must never reach payment. */
export function validateOrder(order: Order, user: User): void {
  if (user.status !== CustomerStatus.Active) {
    throw new OrderValidationError(`customer ${user.id} is ${user.status}`);
  }
  if (order.lines.length === 0) {
    throw new OrderValidationError('order has no lines');
  }
  if (order.total.cents <= 0) {
    throw new OrderValidationError('order total must be positive');
  }
}
